File: marti/Model/VintageModel/VintageDataImporter.m
//
//  VintageDataImporter.m
//  MartiPro
//
//  Created by Marco Guilmette on 2/13/2014.
//  Copyright (c) 2014 Infologique. All rights reserved.
//

#import "VintageDataImporter.h"

#import "VintageMartiCDManager.h"
#import "MartiCDManager.h"

//Vintage Model
#import "ILgroup.h"
#import "ILstep.h"
#import "ILstepAudio.h"
#import "ILstepAudioShort.h"
#import "ILstepImage.h"
#import "ILstepThumbnail.h"
#import "ILstepVideo.h"
#import "ILtask.h"
#import "ILtaskAudio.h"
#import "ILtaskAudioShort.h"
#import "ILtaskImage.h"
#import "ILtasksList.h"
#import "ILtasksListAudio.h"
#import "ILtasksListThumbnail.h"
#import "ILtaskThumbnail.h"

//New Model
#import "Task.h"
#import "MediaStep.h"
#import "RoutingStep.h"
#import "Media+Human.h"
#import "User.h"
#import "Calendar.h"


@implementation VintageDataImporter {
    NSManagedObjectContext* sourceMOC;
    NSManagedObjectContext* destinationMOC;
    
    User *newUser;
}

+ (VintageDataImporter*)sharedInstance
{
    
    static VintageDataImporter *_sharedInstance = nil;
    static dispatch_once_t oncePredicate;
    
    dispatch_once(&oncePredicate, ^{
        _sharedInstance = [[VintageDataImporter alloc] init];
    });
    return _sharedInstance;
}


-(Task*)importVintageTask:(ILtask*)vintageTask{
    
    NSEntityDescription *taskEntity = [NSEntityDescription entityForName:@"Task" inManagedObjectContext:destinationMOC];
    Task* newTask = (Task*)[[NSManagedObject alloc] initWithEntity:taskEntity insertIntoManagedObjectContext:destinationMOC];
    
    newTask.name = vintageTask.task_title;
    newTask.audioAssistant = vintageTask.task_audio.taskAudio_data;
    newTask.index =  vintageTask.task_fakeIndex;
    newTask.mediaAssistant.data =  vintageTask.task_image.taskImage_data;
    //newTask vintageTask.task_language;
    newTask.thumbnail = vintageTask.task_thumbnail.taskThumbnail_data;
    
    newTask.user = newUser;
    
    for( ILstep* vintageStep in vintageTask.steps) {
        //NSLog(@"StepStyle = %@", vintageStep.step_style);
        if( [vintageStep.step_style isEqualToString:@"simpleStep"]  ) {

            [self importVintageSimpleStep:vintageStep].task = newTask;
            
        } else { //choiceStep
            
            [self importVintageTaskList:vintageStep.subTasks].task = newTask;
        }
        
    }
    
    
    
    return newTask;
}


-(Media*)importMediaData:(NSData*)data ofType:(enum MediaType)mediaType{
    
    NSEntityDescription *taskEntity = [NSEntityDescription entityForName:@"Media" inManagedObjectContext:destinationMOC];
    Media* newMedia = (Media*)[[NSManagedObject alloc] initWithEntity:taskEntity insertIntoManagedObjectContext:destinationMOC];
    
    newMedia.type = [NSNumber numberWithInt:mediaType];
    newMedia.data = data;
    
    return newMedia;
}

-(MediaStep*)importVintageSimpleStep:(ILstep*)vintageStep{
    NSEntityDescription *taskEntity = [NSEntityDescription entityForName:@"MediaStep" inManagedObjectContext:destinationMOC];
    MediaStep* newMediaStep = (MediaStep*)[[NSManagedObject alloc] initWithEntity:taskEntity insertIntoManagedObjectContext:destinationMOC];
    
    newMediaStep.name = vintageStep.step_title;
    newMediaStep.textualAssistant = vintageStep.step_description;
    newMediaStep.thumbnail = vintageStep.step_thumbnail.stepThumbnail_data;
    
    //NSLog(@"%@", vintageStep.step_contentStyle);
    
    if ( [vintageStep.step_contentStyle isEqualToString:@"imageAndAudio"] ) {

        newMediaStep.audioAssistant = vintageStep.step_audio.stepAudio_data;
        
        if( vintageStep.step_image.stepImage_data != nil)
        newMediaStep.mediaAssistant = [self importMediaData: vintageStep.step_image.stepImage_data ofType:MediaTypeImage];
        
        NSLog(@"%@ is imageAndAudio", vintageStep.step_title);
        
    }
    
    if ( [vintageStep.step_contentStyle isEqualToString:@"video"] ) {
        
        if( vintageStep.step_video.stepVideo_data != nil)
        newMediaStep.mediaAssistant = [self importMediaData: vintageStep.step_video.stepVideo_data ofType:MediaTypeVideo];
        
        NSLog(@"%@ is video", vintageStep.step_title);
    }
    
    return newMediaStep;
}

-(RoutingStep*)importVintageTaskList:(ILtasksList*)vintageTaskList{
    NSEntityDescription *taskEntity = [NSEntityDescription entityForName:@"RoutingStep" inManagedObjectContext:destinationMOC];
    RoutingStep* newRoutingStep = (RoutingStep*)[[NSManagedObject alloc] initWithEntity:taskEntity insertIntoManagedObjectContext:destinationMOC];
    
    newRoutingStep.name = vintageTaskList.tasksList_title;
    
    //Loop throught all task in this ILtaskList
    for ( ILtask* vintageTask in vintageTaskList.tasks) {
        
        //TODO: Find how to link task to RoutingStep HERE
        //[newRoutingStep addTasksObject:[self importVintageTask:vintageTask]];
        
        Task* newTask = [self importVintageTask:vintageTask];
        [newTask addRoutingStepsObject:newRoutingStep];
        newTask.isRoot = [NSNumber numberWithBool:NO];
     
    }
    
    return newRoutingStep;
}

-(void)importData{
    
    [self clearStore];
    
    sourceMOC = [[VintageMartiCDManager sharedManager] managedObjectContext];
    destinationMOC = [[MartiCDManager sharedInstance] managedObjectContext];
    
    newUser = [NSEntityDescription insertNewObjectForEntityForName:@"User" inManagedObjectContext:destinationMOC];
    newUser.name = @"Default";
    
    //Create the default calendar (could be directly linked to user ???)
    Calendar* newCalendar = [NSEntityDescription insertNewObjectForEntityForName:@"Calendar" inManagedObjectContext:destinationMOC];
    newCalendar.user = newUser;
    
    NSFetchRequest* fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"ILtask"];
    
    NSArray* VintageTasks = [sourceMOC executeFetchRequest:fetchRequest error:nil];
    
    for( ILtask* vintageTask in VintageTasks ) {
        //If this is a root task
        if ( [vintageTask.belongToTasksList.tasksList_type isEqualToString:@"root"] ) {
            [self importVintageTask:vintageTask];
            [[MartiCDManager sharedInstance] saveContext];
            
            NSLog(@"Importing vintage task : %@", vintageTask.task_title);
        }
    }

}

-(void)clearStore{
    [[MartiCDManager sharedInstance] deleteAllObjects:@"User"];
    [[MartiCDManager sharedInstance] deleteAllObjects:@"Media"];
}

@end

File: marti/Model/VintageModelImporter/VintageDataImporter.swift
//
//  VintageDataImporter.swift
//  marti
//
//  Created by Marco Guilmette on 2015-08-19.
//  Copyright (c) 2015 Infologique. All rights reserved.
//

import Foundation
import CoreData

class VintageDataImporter {
    
    enum MediaType {
        case MediaTypeImage
        case MediaTypeVideo
    }
    
    var sourceMOC : NSManagedObjectContext?
    var destinationMOC : NSManagedObjectContext?
    
    init(){
        sourceMOC = VintageCDManager.sharedInstance.managedObjectContext
        destinationMOC = MartiCDManager.sharedInstance.managedObjectContext
    }
    
    func clearMOC(){
        //MartiCDManager.sharedInstance.
    }
    
    func importVintageTaskList(vintageTaskList : ILtasksList) -> RoutingStep {
        let newRoutingStep = NSEntityDescription.insertNewObjectForEntityForName("RoutingStep", inManagedObjectContext: destinationMOC!) as! RoutingStep
        
        if let name = vintageTaskList.tasksList_title { newRoutingStep.name = name }
        //newRoutingStep.name = vintageTaskList.tasksList_title
        
        for vintageTask in vintageTaskList.tasks {
            
            let newTask = importVintageTask(vintageTask as! ILtask)
            
            //[newTask addRoutingStepsObject:newRoutingStep];
            var mutableItems = newRoutingStep.tasks.mutableCopy() as! NSMutableOrderedSet
            mutableItems.addObject(newTask)
            newRoutingStep.tasks = mutableItems.copy() as! NSOrderedSet
            
            newTask.isRoot = true
        }
        
        return newRoutingStep
    }
    

    func importMediaData(data: NSData, mediaType: MediaType) -> Media {
        let newMedia = NSEntityDescription.insertNewObjectForEntityForName("Media", inManagedObjectContext: destinationMOC!) as! Media
        newMedia.type = mediaType.hashValue
        newMedia.data = data
        
        return newMedia
    }

    func importVintageSimpleStep(vintageStep: ILstep) -> MediaStep{
        //Create a new MediaStep entity in new model
        let newMediaStep = NSEntityDescription.insertNewObjectForEntityForName("MediaStep", inManagedObjectContext: destinationMOC!) as! MediaStep
        
        //Thoses value can be nil so we unwrap the optional
        if let name = vintageStep.step_title {newMediaStep.name = name}
        if let description = vintageStep.step_description { newMediaStep.textualAssistant = description }

        newMediaStep.thumbnail = vintageStep.step_thumbnail.stepThumbnail_data
        
        if vintageStep.step_contentStyle == "imageAndAudio" {
            newMediaStep.audioAssistant = vintageStep.step_audio.stepAudio_data
            if vintageStep.step_image.stepImage_data.length == 0 {
                newMediaStep.mediaAssistant = importMediaData(vintageStep.step_image.stepImage_data, mediaType: .MediaTypeImage)
            }
        }
        
        if vintageStep.step_contentStyle == "video" {
            newMediaStep.mediaAssistant = importMediaData(vintageStep.step_video.stepVideo_data, mediaType: .MediaTypeVideo)
        }
        
        return newMediaStep
    }
    
    func importVintageTask(vintageTask : ILtask) -> Task {
        let newTask = NSEntityDescription.insertNewObjectForEntityForName("Task", inManagedObjectContext: destinationMOC!) as! Task
        if let name = vintageTask.task_title { newTask.name = name }
        newTask.audioAssistant = vintageTask.task_audio.taskAudio_data
        newTask.index =  vintageTask.task_fakeIndex
        newTask.mediaAssistant.data =  vintageTask.task_image.taskImage_data
        //newTask vintageTask.task_language;
        newTask.thumbnail = vintageTask.task_thumbnail.taskThumbnail_data
        newTask.user = MartiCDManager.sharedInstance.currentUser

        //TODO: Find out why we are unable to set to a for in loop
        for var index = 0; index < vintageTask.steps.count; index++ {
            var vintageStep = vintageTask.steps[index] as! ILstep
            
            if vintageStep.step_style == "simpleStep" {
                //TODO: Plante quand un attribut est vide (nil).
                importVintageSimpleStep(vintageStep).task = newTask
            } else {
                importVintageTaskList(vintageStep.subTasks).task = newTask
            }
        }
        return newTask
    }
    
    func importData(){
        clearMOC()
        
        var user = MartiCDManager.sharedInstance.currentUser
        
        let fetchRequest = NSFetchRequest(entityName: "ILtask")
        let vintageTasks : [ILtask] = sourceMOC?.executeFetchRequest(fetchRequest, error: nil) as! [ILtask]

        //let vintageTask : ILtask
        for vintageTask in vintageTasks {
            if vintageTask.belongToTasksList.tasksList_type == "root" {
                importVintageTask(vintageTask)
            }
        }
        
        //print("Username = \(user.name)")
       
        MartiCDManager.sharedInstance.saveContext()
    }
    

}